### ISSUE TITLE
BSONDecoder doesn't work with optionals generic values

### ISSUE THREAD

**Issue by valeriomazzeo** [mongodb/mongo-swift-driver#172]
`BSONDecoder` has troubles when trying to decode optional generic values.

Take into account this sample code:

```swift
struct GenericStruct<C: Codable & Equatable>: Codable, Equatable {
  let value: C
}

func testOptionals() throws {
        let encoder = BSONEncoder()
        let decoder = BSONDecoder()

        let c1 = GenericStruct<Int?>(value: 1)
        let c1Doc: Document = ["value": 1]
        XCTAssertEqual(try encoder.encode(c1), c1Doc)
        XCTAssertEqual(try decoder.decode(GenericStruct<Int?>.self, from: c1Doc), c1)

        let c2 = GenericStruct<Int?>(value: nil)
        let c2Doc: Document = ["value": nil]
        XCTAssertEqual(try encoder.encode(c2), c2Doc)
        XCTAssertEqual(try decoder.decode(GenericStruct<Int?>.self, from: c2Doc), c2)
}
```

You'll see that `XCTAssertEqual(try decoder.decode(GenericStruct<Int?>.self, from: c2Doc), c2)` will fail.

Note: I have tested this on `0.0.7`, therefore I am able to `let c2Doc: Document = ["value": nil]`.
On the latest `master` branch it's not possible to set `nil`, and setting `NSNull` instead seems to have fixed the issue. For example: `let c2Doc: Document = ["value": NSNull()]`

Although omitting the key`let c2Doc: Document = [:]` doesn't work on either versions:
```
XCTAssertEqual failed: threw error "keyNotFound(CodingKeys<Optional<Int>>(stringValue: "value", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys<Optional<Int>>(stringValue: \"value\", intValue: nil) (\"value\").", underlyingError: nil))" - 
```

It would be great if this issue could be fixed in all the edge cases and a new version released.

### Swift version
4.2.1

### Operating system

What does this command give you?
18.2.0 Darwin Kernel Version 18.2.0: Mon Nov 12 20:24:46 PST 2018; root:xnu-4903.231.4~2/RELEASE_X86_64 x86_64

### Driver version

```
{
  "object": {
    "pins": [
      {
        "package": "Nimble",
        "repositoryURL": "https://github.com/Quick/Nimble.git",
        "state": {
          "branch": null,
          "revision": "cd6dfb86f496fcd96ce0bc6da962cd936bf41903",
          "version": "7.3.1"
        }
      },
      {
        "package": "bson",
        "repositoryURL": "https://github.com/mongodb/swift-bson",
        "state": {
          "branch": null,
          "revision": "87b458f55ba66f20305341d9d138e47dbcc92da7",
          "version": "2.0.0"
        }
      },
      {
        "package": "mongoc",
        "repositoryURL": "https://github.com/mongodb/swift-mongoc",
        "state": {
          "branch": null,
          "revision": "e503b50210dae2cf0220474c2f12087830150e9f",
          "version": "2.0.0"
        }
      }
    ]
  },
  "version": 1
}
```

### What is the version(s) of `mongod` that you are running with the driver?
3.6.8

**Comment by kmahar** [comment]
Thanks for the detailed report!

So as for the difference between `0.0.7` and `master` with `nil` vs `NSNull`, this was done in e403e2ca26bebc34cd081a3b96b87f071e184741 to disambiguate values that are not present from values that are present and set to `NSNull` (representing the BSON Null type.) However, note that due to some Objective-C baggage we've ended up changing from `NSNull` to a custom `BSONNull` type in #164. Once #164 is merged in we can release these changes.

As for your point about decoding failing when the `Document` is empty, that is expected behavior and is consistent with `JSONDecoder`. Using your type from above:
```swift
let emptyData = "{}".data(using: .utf8)!
XCTAssertEqual(try JSONDecoder().decode(GenericStruct<Int?>.self, from: emptyData), c2)
```
This gives ` XCTAssertEqual failed: threw error "keyNotFound(CodingKeys<Optional<Int>>(stringValue: "value", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys<Optional<Int>>(stringValue: \"value\", intValue: nil) (\"value\").", underlyingError: nil))"`.

It works when the key is present in the JSON and set to `null`:
```
let data = "{\"value\": null}".data(using: .utf8)!
XCTAssertEqual(try JSONDecoder().decode(GenericStruct<Int?>.self, from: data), c2)
```

We'd like to keep our behavior consistent with `JSONEncoder` and `JSONDecoder`, so it is unlikely we will make changes to allow optionals to be decoded as missing keys. Perhaps you can find a workaround by writing a custom `init(from decoder: Decoder)` method?

**Comment by valeriomazzeo** [comment]
I think you are right, and I agree it should behave exactly as `JSONDecoder`.

From what I understand the different behaviour (both in `BSONDecoder` and `JSONDecoder` is due to how Swift automatically synthesise `init(from decoder: Decoder) throws` from that struct.

I suspect that when a type is specifically declared as `Optional` it uses something like:
```swift
self.value = try container.decodeIfPresent(C.self, forKey: .value)`
```
Instead when it is declared as `C` (like in the example) where `C` could be `Int?`, it gets synthesised as:
```swift
self.value = try container.decode(C?.self, forKey: .value)`
```
Which cause the different behaviour between a missing key and `null`.

Having said that, I don't have control over this type I am working with, so I will have to deal with it.

Although, it seems that once the latest changes are going to be released, I should be able to ensure the key is present containing `NSNull`, and therefore I can have a way forward.

Thanks for taking the time to read these issues, I really appreciate it.

**Comment by kmahar** [comment]
Aha - yes, that makes perfect sense about the synthesized code. Interesting how generics and optionals interact here. 

I agree that normalizing your data to always have the field set to `BSONNull` seems like the best workaround here. We should be able to release the changes shortly, they are in review now. 

Thanks so much for opening issues, you're helping us make the driver better! Out of curiosity, could you share a bit more info about how you're using it?

**Comment by valeriomazzeo** [comment]
I am working on a driver for `Fluent` (Vapor ORM) https://github.com/asensei/vapor-fluent-mongo to use with our Swift backend.

**Comment by kmahar** [comment]
Sorry, forgot to reply here - awesome! Followed the repo, will be interested to see how it comes together. 

Closing this issue as we determined the problem is with synthesized `Codable` conformance and the `BSONNull` type allowing a workaround was implemented in 38665745a20eec58b2797f86e4ae6bfac23e5fb1.
